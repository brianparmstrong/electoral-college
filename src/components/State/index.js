import React, { useEffect, useRef, useState } from 'react';
import { bool, func, number, oneOfType, string } from 'prop-types';
import { getWinnerClassName } from '../../utilities';

const State = ({
  evs,
  isFromStorage,
  name,
  stateClickedFromMap,
  stateCode,
  stateEvs,
  toggleWinner,
  winner
}) => {
  const stateNode = useRef(null);
  const [winningParty, setWinningParty] = useState(
    isFromStorage ?
    Number(winner) :
    0
  );
  const [winningPartyClass, setWinningPartyClass] = useState(getWinnerClassName(winner));

  useEffect(() => {
    stateClickedFromMap && stateNode.current.click();
  }, [ stateClickedFromMap ]);

  const updateWinningParty = e => {
    const newWinningParty = (winningParty === 5) ? 0 : (winningParty + 1);
    let newWinningPartyClass = '';
    let newPreviousWinningPartyClass = '';

    switch (newWinningParty) {
      case 1:
        newWinningPartyClass = 'gop';
        newPreviousWinningPartyClass = '';
        break;
      case 2:
        newWinningPartyClass = 'dem';
        newPreviousWinningPartyClass = 'gop';
        break;
      case 3:
        newWinningPartyClass = 'lib';
        newPreviousWinningPartyClass = 'dem';
        break;
      case 4:
        newWinningPartyClass = 'grn';
        newPreviousWinningPartyClass = 'lib';
        break;
      case 5:
        newWinningPartyClass = 'ind';
        newPreviousWinningPartyClass = 'grn';
        break;
      default:
        newWinningPartyClass = '';
        newPreviousWinningPartyClass = 'ind';
        break;
    }

    const newData = {
      evs: parseInt(evs),
      newPreviousWinnerTargetElem: `${newPreviousWinningPartyClass}-wta`,
      newWinningParty,
      newWinningTargetElem: `${newWinningPartyClass}-wta`,
      stateId: e.target.getAttribute('data-statecode')
    };

    setWinningParty(newWinningParty);
    setWinningPartyClass(newWinningPartyClass);
    toggleWinner(newData);
  };

  return (
    <div className='stateWrapper' id={name}>
      <div 
        className={`state ${winningPartyClass}`}
        data-evs={evs}
        data-stateevs={stateEvs}
        data-fromstorage={isFromStorage}
        data-statecode={stateCode}
        data-statename={name}
        data-winner={winningParty}
        onClick={updateWinningParty}
        ref={stateNode}
        tabIndex='0' >
        <span className='evs'>{evs}</span>
      </div>
      <p className='stateName'>{name}</p>
    </div>
  );
};

State.propTypes = {
  evs: oneOfType([number, string]).isRequired,
  isFromStorage: bool.isRequired,
  name: string.isRequired,
  stateClickedFromMap: oneOfType([ bool, string]),
  stateCode: string.isRequired,
  stateEvs: number,
  toggleWinner: func.isRequired,
  winner: string
};

export default State;
