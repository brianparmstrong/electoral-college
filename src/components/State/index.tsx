import { MouseEvent, useEffect, useRef, useState } from 'react';
import { IfcState } from '../../types';
import { getWinnerClassName } from '../../utilities';

const State = (stateInfo: IfcState) => {
  const {
    evs,
    isFromStorage,
    name,
    stateClickedFromMap,
    stateCode,
    stateEvs,
    toggleWinner,
    winner,
  } = stateInfo;
  const stateNode = useRef(null);
  const [winningParty, setWinningParty] = useState(
    isFromStorage ? Number(winner) : 0
  );
  const [winningPartyClass, setWinningPartyClass] = useState(
    getWinnerClassName(winner)
  );

  useEffect(() => {
    stateClickedFromMap !== 'false' && stateNode.current.click();
  }, [stateClickedFromMap]);

  const updateWinningParty = (event: MouseEvent<HTMLDivElement>) => {
    const newWinningParty = winningParty === 5 ? 0 : winningParty + 1;
    let newWinningPartyClass = '';

    switch (newWinningParty) {
      case 1:
        newWinningPartyClass = 'gop';
        break;
      case 2:
        newWinningPartyClass = 'dem';
        break;
      case 3:
        newWinningPartyClass = 'lib';
        break;
      case 4:
        newWinningPartyClass = 'grn';
        break;
      case 5:
        newWinningPartyClass = 'ind';
        break;
      default:
        newWinningPartyClass = '';
        break;
    }

    const newData = {
      evs: parseInt(evs),
      newWinningParty: String(newWinningParty),
      stateId: event.currentTarget.dataset.statecode,
    };

    setWinningParty(newWinningParty);
    setWinningPartyClass(newWinningPartyClass);
    toggleWinner(newData);
  };

  return (
    <div className="stateWrapper" id={name}>
      <div
        className={`state ${winningPartyClass}`}
        data-evs={evs}
        data-stateevs={stateEvs}
        data-fromstorage={isFromStorage}
        data-statecode={stateCode}
        data-statename={name}
        data-winner={winningParty}
        onClick={updateWinningParty}
        ref={stateNode}
        tabIndex={0}
      >
        <span className="evs">{evs}</span>
      </div>
      <p className="stateName">{name}</p>
    </div>
  );
};

export default State;
