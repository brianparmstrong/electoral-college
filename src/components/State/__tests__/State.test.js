import { render } from '@testing-library/react';
import State from '../';

const MOCK_STATE_PROPS = {
  evs: 15,
  isFromStorage: false,
  name: 'North Carolina',
  stateClickedFromMap: true,
  stateCode: 'NC',
  toggleWinner: jest.fn(),
  winner: 3
};

it('renders correctly', () => {
  const { container } = render(<State { ...MOCK_STATE_PROPS } />);
  expect(container.firstChild).toMatchSnapshot();
});
