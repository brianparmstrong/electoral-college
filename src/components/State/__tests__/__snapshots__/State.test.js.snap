// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="stateWrapper"
  id="North Carolina"
>
  <div
    className="state lib"
    data-evs={15}
    data-fromstorage={false}
    data-statecode="NC"
    data-statename="North Carolina"
    data-winner={0}
    onClick={[Function]}
    tabIndex="0"
  >
    <span
      className="evs"
    >
      15
    </span>
  </div>
  <p
    className="stateName"
  >
    North Carolina
  </p>
</div>
`;
